// Copyright 2020 Google Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
////////////////////////////////////////////////////////////////////////////////

// This protobuf represents an X.509 certificate chain with a set of operations
// to mutate the chain to create structural relationships between the
// certificates.

syntax = "proto2";

import "x509_certificate.proto";

package mutated_x509_chain;

message MutatedX509Chain {
  required X509Chain x509_chain = 1;
  repeated Operation operations = 2;
}

// An X.509 Certificate Chain comprises a list of |certificates|.
message X509Chain {
  repeated x509_certificate.X509Certificate certificates = 1;
}

message Operation {
  oneof types {
    MutateSignatureOperation mutate_signature_operation = 1;
  }
  // |index| represents the position of the certificate that will be mutated.
  required CertIndex index = 2;
}

message MutateSignatureOperation {
  // Allow certificate to be valid or invalid for fuzzing.
  required bool is_valid = 1;
}

// Contains the positions of the certiciate chain that will be mutated.
enum CertIndex {
  pos_0 = 0;
  pos_1 = 1;
  pos_2 = 2;
  pos_3 = 3;
  pos_4 = 4;
}